Az357 eunit

slfritchie opened the pull request:
Most of this branch is already merged into master, per earlier agreement with Bryan.
Review is a mostly pro forma thing.

slfritchie committed:
Revert a few patches after GitHub & phone review

slfritchie committed:
Add pipeline ordering test

slfritchie committed:
Add straggler process test

slfritchie committed:
Add worker init failure tests

slfritchie committed:
Add trace filtering test

slfritchie committed:
Add worker limit enforcement tests

slfritchie committed:
Add worker queue limit enforcement test

beerriot commented:
The startup/teardown stuff is awesome.  It might be nice to export those functions when `TEST` is defined, such that they can be used from an erl shell.

If I understand these tests correctly, I think they're targeting the correct points in the system.  I found the `generic_transform` tests a bit hard to follow, because the crash depends not only on the behavior of `DecrOrCrashFun`, specified in one place, and the `DriverFun` (`XBad`, `…Crash`, …), specified in another, but also on the length of the pipeline created, specified in a third place in the file.  Maybe a little English next to each test case could help explain why the test is expected to crash in the manner it crashes?

I also solved the problem I was having running the tests earlier:  if you use an old riak_core version, without the `handle_info/2` function in the `riak_core_vnode` behavior, vnodes never get `'DOWN'` messages from fittings, so they happily keep queueing inputs for pipelines that have vanished after their fittings crash.  So, if you see `generic_transform(TailFittingCrash)`, `generic_transform(MiddleFittingCrash)`, `generic_transform(MiddleFittingNormal)`, and `generic_transform(HeadFittingCrash)` fail, try updating your riak_core dependency.

Finally, this is a significant amount of test code, and I wonder if we should move it out of `riak_pipe.erl` and into a `test/riak_pipe_test.erl` (or several), just to separate things a bit.

+1 to merge, otherwise.

slfritchie commented:
I'm generally not a proponent of putting tests code into a "test" subdir, so I'll be lazy.

The suggestion about commenting some of the exception test funs is good, I'll fiddle with those and then merge.

slfritchie committed:
Moral of the testing story: don't be too afraid of indentation.

I'd put many of the critical/per-test-different bits into anonymous
functions that were defined too far away from where they were actually
used.  Rationale: less worries about indentation & max 80 character lines.
In reality, it was too difficult to read, and after trying to move a few
functions, the 80 char line limit was no problem.
